### PR TITLE
[tests-only][full-ci] Test scenarios for relative quota amount of personal space

### PR DESCRIPTION
**Pull request by roshanlc**
## Description
This PR adds test scenario for confirming the relative quota amount of personal space.

## Related Issue
- Related to https://github.com/owncloud/ocis/issues/4357

## How Has This Been Tested?

- Locally
- CI

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 1]
```diff
@@ -3134,4 +3134,52 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+			$key = $row['key'];
+			$value = $row['value'];
+			$segments = explode('@@@', $key);
+			if (\sizeof($segments) > 1) {
+				$arrayKeyExists = \array_key_exists($segments[0], $data);
+				Assert::assertTrue($arrayKeyExists, "The key $key does not exist on the response");
+				$final_key_value  = $data[$segments[0]];
+				$last_array =  $data[$segments[0]];
+				// Move inside nested blocks each step at a time
+				for ($i = 1;$i < \sizeof($segments);$i++) {
```
Comment (amrita-shrestha) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3163]:
```suggestion
				for ($i = 1;$i < \sizeof($segments);$i++) {
```
for variable `$i` give some suitable name

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 2]
```diff
@@ -3134,4 +3134,52 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+	 *
+	 * @param TableNode $table
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function theJsonRespondedShouldContainTheseKeyAndValuePairs(TableNode $table):void {
+		$this->featureContext->verifyTableNodeColumns($table, ['key', 'value']);
+		$response = json_decode((string)$this->featureContext->getResponse()->getBody(), true, 512, JSON_THROW_ON_ERROR);
+		$data = $response["ocs"]["data"];
```
[the remark this answers is not in the document]
Comment (roshanlc) [reply tests/acceptance/features/bootstrap/SpacesContext.php:3152]: Could you explain more?

File: tests/acceptance/features/apiSpaces/quota.feature [hunk 3]
```diff
@@ -65,3 +65,24 @@ Feature: State of the quota
     And user "Alice" has uploaded a file inside space "Project Delta" with content "7 bytes" to "test.txt"
     When user "Alice" uploads a file inside space "Project Delta" with content "00011 bytes" to "test.txt" using the WebDAV API
     Then the HTTP status code should be "507"
+
+
+  Scenario Outline: Confirm the relative amount of quota of personal space
+    Given user "Admin" has changed the quota of the personal space of "Alice Hansen" space to "10000"
+    And user "Alice" has uploaded a file from "<file_upload>" to "uploaded.txt" via TUS inside of the space "Alice Hansen" using the WebDAV API
```
Comment (SwikritiT) [inline tests/acceptance/features/apiSpaces/quota.feature:72]: why TUS upload, can't we do the normal upload? The issue mentions normal upload

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 4]
```diff
@@ -3134,4 +3134,35 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+			$actualValue  = null;
+			$tempArray =  $data;
+			foreach ($segments as $segment) {
+				$arrayKeyExists = \array_key_exists($segment, $tempArray);
+				Assert::assertTrue($arrayKeyExists, "The key $segment does not exist on the response");
+				$actualValue = $tempArray[$segment];
+				$tempArray = $actualValue;
+			}
+			Assert::assertEquals($expectedValue, $actualValue, "Expected $expectedValue but received $actualValue");
+		}
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3166]:
```suggestion
			$actualValue  = $data;
			foreach ($segments as $segment) {
				$arrayKeyExists = \array_key_exists($segment, $tempArray);
				Assert::assertTrue($arrayKeyExists, "The key $segment does not exist on the response");
				$actualValue = $actualValue[$segment];
			}
			Assert::assertEquals($expectedValue, $actualValue, "Expected $expectedValue but received $actualValue");
		}
```
can you check if like this works for the loop

File: tests/acceptance/features/apiSpaces/quota.feature [hunk 5]
```diff
@@ -65,3 +65,24 @@ Feature: State of the quota
...
+    When the user "Alice" requests these endpoints with "GET" with basic auth
+      | endpoint    |
+      | <end_point> |
+    Then the HTTP status code should be "200"
+    And the json responded should contain these key and value pairs:
+      | key              | value            |
+      | quota@@@free     | <quota_free>     |
+      | quota@@@used     | <quota_used>     |
+      | quota@@@total    | <quota_total>    |
+      | quota@@@relative | <quota_relative> |
```
Comment (grgprarup) [inline tests/acceptance/features/apiSpaces/quota.feature:82]: Can't we use the existing steps? `Then the json responded should contain a space "Alice Hansen" with these key and value pairs:`

Comment (saw-jan) [reply tests/acceptance/features/apiSpaces/quota.feature:82]: @roshanlc existing step can be utilized if already there.

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 6]
```diff
@@ -3134,4 +3134,35 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+	 *
+	 * Loops through data items in response to find given key and value pairs
+	 * It does not support field value replacing
+	 *
+	 * @param TableNode $table
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function theJsonRespondedShouldContainTheseKeyAndValuePairs(TableNode $table):void {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3149]: I think we do have the very similar step already

File: tests/acceptance/features/apiSpaces/quota.feature (hunk 5 above)
Comment (roshanlc) [reply tests/acceptance/features/apiSpaces/quota.feature:82]: That method cannot be used as it returns the following response
```json
{
  "ocs": {
    "meta": {
      "status": "ok",
      "statuscode": 100,
      "message": "OK"
    },
    "data": {
      "enabled": "true",
      "quota": {
        "free": 50000000,
        "used": 50000000,
        "total": 100000000,
        "relative": 50
      },
      "email": "alice@example.org",
      "displayname": "Alice Hansen",
      "user-type": "primary",
      "two_factor_auth_enabled": false,
      "last_login": 0
    }
  }
}
```

Is there any pre-existing method for such responses?

File: tests/acceptance/features/bootstrap/SpacesContext.php (hunk 4 above)
Comment (roshanlc) [reply tests/acceptance/features/bootstrap/SpacesContext.php:3166]: Works.

File: tests/acceptance/features/apiSpaces/quota.feature (hunk 5 above)
Comment (saw-jan) [reply tests/acceptance/features/apiSpaces/quota.feature:82]: Ohh! then you need to do `ocs@@@data@@@quota@@@free`

Comment (saw-jan) [reply tests/acceptance/features/apiSpaces/quota.feature:82]: Need new step

File: tests/acceptance/features/bootstrap/SpacesContext.php (hunk 6 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/SpacesContext.php:3149]: looks like there are none

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 7]
```diff
@@ -3134,4 +3134,33 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+		$this->featureContext->verifyTableNodeColumns($table, ['key', 'value']);
+		$response = json_decode((string)$this->featureContext->getResponse()->getBody(), true, 512, JSON_THROW_ON_ERROR);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3151]: we do have function for that
```suggestion
		$response = $this->featureContext->getJsonDecodedResponse($this->featureContext->getResponse())
```

File: tests/acceptance/features/apiSpaces/quota.feature [hunk 8]
```diff
@@ -65,3 +65,24 @@ Feature: State of the quota
...
+    Examples:
+      | file_upload                   | end_point                                 | quota_free | quota_used | quota_total | quota_relative |
+      | /filesForUpload/lorem.txt     | /ocs/v1.php/cloud/users/Alice?format=json | 9301       | 699        | 10000       | 6.99           |
```
Comment (saw-jan) [inline tests/acceptance/features/apiSpaces/quota.feature:85]: Could do similar to the following without specifying `format` query.
```suggestion
      | /filesForUpload/lorem.txt     | /ocs/v1.php/cloud/users/Alice| 9301       | 699        | 10000       | 6.99           |
```
https://github.com/owncloud/ocis/blob/486453b4c91deec3f91a501a4ff4450c2da0e1d6/tests/acceptance/features/bootstrap/OCSContext.php#L933-L941

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 9]
```diff
@@ -3134,4 +3134,27 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+
+	/**
+	 * @Then the relative quota amount should be :quota_amount
+	 *
+	 * @param string $quotaAmount
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function theRelativeQuotaAmountShouldBe($quotaAmount) {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3146]:
```suggestion
	public function theRelativeQuotaAmountShouldBe($quotaAmount): void {
```

File: tests/acceptance/features/bootstrap/SpacesContext.php [hunk 10]
```diff
@@ -3134,4 +3134,27 @@ public function publicDownloadsTheFolderFromTheLastCreatedPublicLink(string $res
...
+		$data = $this->ocsContext->getOCSResponseData($this->featureContext->getResponse());
+		if (isset($data->quota, $data->quota->relative)) {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/SpacesContext.php:3148]:
```suggestion
		if (\isset($data->quota, $data->quota->relative)) {
```

Comment (roshanlc) [reply tests/acceptance/features/bootstrap/SpacesContext.php:3148]: Could not find \isset used anywhere else.

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/SpacesContext.php:3148]: yeah, we should have used that. it separates the namespace scope and prevents the functions clash (may not happen very often though)

**Review by saw-jan — COMMENTED** [review]
:+1:

File: tests/acceptance/features/apiSpaces/quota.feature [hunk 11]
```diff
@@ -65,3 +65,19 @@ Feature: State of the quota
...
     Then the HTTP status code should be "507"
+
+
+  Scenario Outline: Check the relative amount of quota of personal space
+    Given user "Admin" has changed the quota of the personal space of "Alice Hansen" space to "10000"
+    And user "Alice" has uploaded file "<file_upload>" to "/demo.txt"
+    When the user "Alice" requests these endpoints with "GET" with basic auth
+      | endpoint    |
+      | <end_point> |
+    Then the HTTP status code should be "200"
```
Comment (SagarGi) [inline tests/acceptance/features/apiSpaces/quota.feature:76]: can we also check `ocs` status code if possible?

File: tests/acceptance/features/bootstrap/OCSContext.php [hunk 12]
```diff
@@ -940,6 +940,24 @@ public function getOCSResponseStatusCode(ResponseInterface $response):string {
 		);
 	}
 
+	/**
+	 * Parses the xml answer to return data items from ocs response
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/OCSContext.php:944]:
```suggestion
```

I think the function name says it all.

File: tests/acceptance/features/apiSpaces/quota.feature (hunk 11 above)
Comment (saw-jan) [reply tests/acceptance/features/apiSpaces/quota.feature:76]: yeah, should be possible with `Then the OCS status code should be "<code>"`

File: tests/acceptance/features/bootstrap/OCSContext.php (hunk 12 above)
Comment (roshanlc) [reply tests/acceptance/features/bootstrap/OCSContext.php:944]: I put the comment just in case people don't  confuse data items for the whole XML DATA

**Review by SagarGi — COMMENTED** [review]
LGTM :+1: if CI is happy

**Review by SwikritiT — APPROVED** [review]

**Review by phil-davis — APPROVED** [review]

**Pull request merged**